pkg/apply/applier.go

```diff
@@ -142,18 +154,130 @@ func (a *Applier) newStatusPoller() (poller.Poller, error) {
+		Flatten().
+		Do()
+	if err := r.Err(); err != nil {
+		return nil, err
+	}
+
+	var infos []*resource.Info
+	err := r.Visit(func(info *resource.Info, err error) error {
+		infos = append(infos, info)
+		return nil
+	})
+	return infos, err
+}
+
+// setNamespaces verifies that
+func (a *Applier) setNamespaces(infos []*resource.Info, defaultNamespace string) error {
+	var crdInfos []*resource.Info
+
+	// find any crds in the set of resources.
+	for _, inf := range infos {
+		if solver.IsCRD(inf) {
+			crdInfos = append(crdInfos, inf)
+		}
+	}
+
+	mapper, err := a.factory.ToRESTMapper()
+	if err != nil {
+		return err
+	}
+	for _, inf := range infos {
+		accessor, _ := meta.Accessor(inf.Object)
+		// if the resource already has the namespace set, we don't
+		// need to do anything
+		if accessor.GetNamespace() != "" {
+			continue
+		}
+
+		gk := inf.Object.GetObjectKind().GroupVersionKind().GroupKind()
+		mapping, err := mapper.RESTMapping(gk)
+
+		if err != nil && !meta.IsNoMatchError(err) {
+			return err
+		}
+
+		if err == nil {
+			// If we find a mapping for the resource type in the RESTMapper,
+			// we just use it.
+			if mapping.Scope == meta.RESTScopeNamespace {
+				inf.Namespace = defaultNamespace
+				accessor.SetNamespace(defaultNamespace)
+			}
+			continue
+		}
+
+		// If we get here, it means the resource does not have the namespace
+		// set and we didn't find the resource type in the RESTMapper. As
+		// a last try, we look at all the CRDS that are part of the set and
+		// see if we get a match on the resource type. If so, we can determine
+		// from the CRD whether the resource type is cluster-scoped or
+		// namespace-scoped. If it is the latter, we set the namespace
+		// to the provided default.
+		var scope string
+		for _, crdInf := range crdInfos {
+			u, _ := crdInf.Object.(*unstructured.Unstructured)
+			group, _, _ := unstructured.NestedString(u.Object, "spec", "group")
+			kind, _, _ := unstructured.NestedString(u.Object, "spec", "names", "kind")
+			if gk.Kind == kind && gk.Group == group {
+				scope, _, _ = unstructured.NestedString(u.Object, "spec", "scope")
+			}
+		}
+
+		switch scope {
+		case "":
+			return fmt.Errorf("can't find scope for resource %s %s", gk.String(), accessor.GetName())
+		case "Cluster":
+			continue
+		case "Namespaced":
+			inf.Namespace = defaultNamespace
+			accessor.SetNamespace(defaultNamespace)
+		}
+	}
+
+	return nil
+}
+
 // readAndPrepareObjects reads the resources that should be applied,
-// handles ordering of resources and sets up the inventory object
-// based on the provided inventory object template.
+// handles ordering of resources and sets up the grouping object
+// based on the provided grouping object template.
 func (a *Applier) readAndPrepareObjects() (*ResourceObjects, error) {
-	infos, err := a.ApplyOptions.GetObjects()
+	infos, err := a.manifestReaderFunc()
 	if err != nil {
 		return nil, err
 	}
+
+	err = a.setNamespaces(infos, a.ApplyOptions.Namespace)
```

Comment:
I believe ApplyOptions.Namespace is only set using the "-n" flag on the command line or "namespace" in the kubeconfig file. I don't think its usually set. If any of these two are set, then we also have to look at "EnforceNamespace" field in the builder. We should probably write some tests for this.

Reply:
You are correct. I have updated the code to also verify that the namespace matches `ApplyOptions.Namespace` for resources that already have a namespace if `EnforceNamespace` is true. Also added tests for this.